packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js

```diff
@@ -55,49 +26,80 @@ export const useMagentoRoute = props => {
+            }
+        },
+        [setComponentMap]
+    );
 
-    // ask Magento for a RootComponent that matches the current pathname
-    useEffect(() => {
-        // Avoid setting state if unmounted.
-        if (!isMountedRef.current) {
-            return;
-        }
+    const storeCodeResult = useQuery(getStoreCodeQuery, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first'
+    });
+
+    const urlResult = useQuery(resolveUrlQuery, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first',
+        variables: { url: pathname }
+    });
 
-        if (shouldFetch(routeData, store)) {
-            getRouteComponent(apiBase, pathname, store).then(
-                ({
-                    component,
-                    id,
-                    pathname,
-                    redirectCode,
-                    relativeUrl,
-                    routeError,
-                    type
-                }) => {
-                    // Update our Map in local state for this path.
-                    setComponentMap(prevMap => {
-                        const nextMap = new Map(prevMap);
-
-                        const nextValue = routeError
-                            ? talonResponses.ERROR(routeError)
-                            : id === -1
-                            ? talonResponses.NOT_FOUND
-                            : REDIRECT_CODES.includes(redirectCode)
-                            ? talonResponses.REDIRECT(relativeUrl)
-                            : talonResponses.FOUND(component, id, type, store);
-
-                        return nextMap.set(pathname, nextValue);
-                    });
-                }
-            );
+    // destructure store code result
+    const { data: storeCodeData, error: storeCodeError } = storeCodeResult;
+    const { storeConfig } = storeCodeData || {};
+    const { code: store } = storeConfig || {};
+    const routeKey = getRouteKey(pathname, store);
+
+    // destructure url result
+    const { data: urlData, error: urlError } = urlResult;
+    const { urlResolver } = urlData || {};
+    const { id, redirectCode, relative_url, type } = urlResolver || {};
+
+    // calculate response
+    const component = componentMap.get(routeKey);
+    const isEmpty = !store || !urlResolver || !type || id < 1;
+    const isPending = storeCodeResult.loading || urlResult.loading;
+    const isRedirect = REDIRECT_CODES.has(redirectCode);
+    const fetchError = component instanceof Error && component;
+    const queryError = fetchError || storeCodeError || urlError;
+
+    const routeData =
+        component && !fetchError
+            ? RESPONSES.FOUND(component)
+            : queryError
+            ? RESPONSES.ERROR(queryError)
+            : isEmpty && !isPending
+            ? RESPONSES.NOT_FOUND
+            : isRedirect
+            ? RESPONSES.REDIRECT(relative_url)
+            : RESPONSES.LOADING;
```

Comment:
Is it just me or are these really hard to read?

Reply:
I always do a double-take on tenaries that aren't simply two values. We usually rewrite them as overblown if/elseif/else trees, so it's super obvious in spite of syntax characters making it "cluttered".